migration-to-python-3-only.md

```diff
@@ -0,0 +1,207 @@
+through `pip install ipython` or equivalent under python 2, will end up with a
+broken installation.
+
+The we need to to fix at least the following issues for this to work:
+
+    https://github.com/pypa/pip/issues/3182
+
+Note that this will not be an issue before we actually release a source version
+of a package which is not Python 2 compatible, but that we likely want to have
+this fixed in pip as soon as possible so that users have a chance to upgrade to
+this version **before** upgrading IPython.
+
+There are alternative ways around that, like have IPython being only a meta package, 
+
+
+
+
+
+# Notebook & Python 3 support. 
+
+The way the Jupyter Protocol is designed, it is language agnostic, thus it is
+perfectly possible to imagine that the notebook server (in particular), could
+be written in a non-python language. 
+
+For the next+1 version of the Notebook (6.0), we propose to require a Python 3 environment. 
+
+The rational being that the notebook application is one of those that can
+benefit the most from recent Python improvement, being asyncio, `yiield from`, etc. 
+
+Unlike for IPython, we do not believe a LTS support is necessary, as a Python 3
+only version of the notebook application can perfectly run a Python 2 kernel.
+Thus a Python 3 only version of the notebook is technically not really dropping Python 2.
+
+In order to allow progressive transition to Python 3 for users and downstream
+integrators, we propose to release the first Python 3 only version of the
+notebook 6 month after the first release of the Python 3 only version of IPython. 
+
+We propose to strongly communicate for next release of the notebook (5.0), that
+one of the next version will require a Python 3 environment. 
+
+Timelines:
+    - Notebook 5.0 – Python 2 compatible – Summer/Fall 2016
+    - IPython 6.0 – Python 3 Only – Early 2017
+    – Notebook 6.0 (or 7.0) – Python 3 only – Mid 2017.
+
+
+
+
+## Challenges
+
+As for IPython above; the notebook package need to be able to express that it
+requires Python 3. This is mitigated by people installing `jupyter` that we can
+make conditionally depend on different version of the notebook depending on the
+version of Python. 
+
+In particular for Anaconda, one of the path forward would be to bundle the
+notebook with its own python 3 interpreter. This can lead to small
+inconsistency for notebook server extensions, as on Python 2 Anaconda, the
+notebook server will be in a separate environment, and the notebook extensions
+and optional dependencies need to be installed in this environment as well. We
+are thus speaking of `nbconvert`, custom exporters, custom handler and server
+side extension.
+
+The anaconda crew told us that dropping support for Python 2 will lead them to
+unbundle Jupyter from Anaconda instead of using the 2 environment trick.
+
+This is not a problem on linux and unix-like system where notebook will "just" requires Python 3. 
```

Comment:
This isn't very clear: it's not the unixy-ness that matters here, it's how different packaging systems have chosen to handle Python 2 and 3. It's still a problem for Anaconda on Unixy systems, but it's not a problem for apt/rpm, which treat Python 2 and 3 as two different things.

(Of course, it's anyone's guess when, if ever, Jupyter will be packaged in Debian. But this is one thing that won't be a problem for them.)